File: frontend/src/js/student.js
// -------------------- Load Logged-In User --------------------
const user = JSON.parse(localStorage.getItem("user"));
if (!user) window.location.href = "./login.html";  // Redirect if not authenticated.

const welcome = document.getElementById("welcome");
const logout = document.getElementById("logout");
const tableBody = document.getElementById("courseRows");
const actionHeader = document.getElementById("actionHeader");

// Display greeting message.
welcome.textContent = `Welcome ${user.name}!`;
let showingMyCourses = true;


// -------------------- Logout Listener --------------------
logout.addEventListener("click", () => {
  localStorage.removeItem("user");  // Clear session info.
  window.location.href = "./login.html";
});


// -------------------- Tab Switching --------------------
document.getElementById("myCoursesTab").addEventListener("click", () => {
  showingMyCourses = true;
  loadMyCourses();  // Load only the student's courses.
  document.getElementById("myCoursesTab").classList.add("active");
  document.getElementById("allCoursesTab").classList.remove("active");
});

document.getElementById("allCoursesTab").addEventListener("click", () => {
  showingMyCourses = false;
  loadAllCourses();  // Load all available courses.
  document.getElementById("allCoursesTab").classList.add("active");
  document.getElementById("myCoursesTab").classList.remove("active");
});


// -------------------- Load Student's Enrolled Courses --------------------
async function loadMyCourses() {
  actionHeader.textContent = "Drop";  // Column header changes depending on mode.

  const res = await fetch(`http://127.0.0.1:5000/api/student/myclasses/${user.id}`);
  const data = await res.json();
  tableBody.innerHTML = "";  // Clear previous table contents.

  data.forEach((c) => {
    const row = document.createElement("tr");

    row.innerHTML = `
      <td>${c.class}</td>
      <td>${c.teacher}</td>
      <td>${c.time}</td>
      <td>${c.students}/${c.capacity}</td>
      <td style="text-align:center; color:red; font-size:20px; cursor:pointer;">−</td>
    `;

    // Drop course handler
    row.lastElementChild.addEventListener("click", async () => {
      if (!confirm(`Are you sure you want to drop ${c.class}?`)) return;

      const res = await fetch("http://127.0.0.1:5000/api/student/drop", {
        method: "DELETE",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ student_id: user.id, class_id: c.id }),
      });

      if (res.ok) {
        alert(`Dropped ${c.class}.`);
        loadMyCourses();  // Refresh list after dropping.
      } else {
        alert("Failed to drop course.");
      }
    });

    tableBody.appendChild(row);
  });
}


// -------------------- Load All Courses --------------------
async function loadAllCourses() {
  actionHeader.textContent = "Add class";

  const res = await fetch("http://127.0.0.1:5000/api/student/classes");
  const all = await res.json();

  // Get the student's current enrollments to mark enrollment state.
  const myRes = await fetch(`http://127.0.0.1:5000/api/student/myclasses/${user.id}`);
  const my = await myRes.json();
  const myNames = my.map((c) => c.class);

  tableBody.innerHTML = "";

  all.forEach((c) => {
    const isFull = c.students >= c.capacity;  // Check if class is full.
    const isEnrolled = myNames.includes(c.name);  // Student already enrolled?
    const icon = isEnrolled ? "−" : isFull ? "✖" : "+";  // UI icon depending on state.
    const color = isEnrolled ? "red" : isFull ? "gray" : "green";

    const row = document.createElement("tr");

    row.innerHTML = `
      <td>${c.name}</td>
      <td>${c.teacher}</td>
      <td>${c.time}</td>
      <td>${c.students}/${c.capacity}</td>
      <td style="text-align:center; color:${color}; font-size:20px; cursor:pointer;">${icon}</td>
    `;

    // Enrollment handler (only when class is available)
    if (!isFull && !isEnrolled) {
      row.lastElementChild.addEventListener("click", async () => {
        const res = await fetch("http://127.0.0.1:5000/api/student/signup", {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({ student_id: user.id, class_id: c.id }),
        });

        if (res.ok) {
          alert(`Successfully enrolled in ${c.name}!`);
          loadAllCourses();  // Refresh display
        } else {
          alert("Failed to enroll.");
        }
      });
    }

    tableBody.appendChild(row);
  });
}


// -------------------- Load “My Courses” by Default --------------------
loadMyCourses();

File: backend/routes/student.py
from flask import Blueprint, jsonify, request
from backend import db
from backend.models.user import User
from backend.models.course import Course
from backend.models.enrollment import Enrollment

# Blueprint containing routes for student-related operations.
student_bp = Blueprint("student", __name__)


# -------------------- Get All Available Classes --------------------
@student_bp.route("/classes", methods=["GET"])
def get_all_classes():
    """
    Returns a list of all courses in the system.
    Each entry includes course details, teacher name,
    current enrollment count, and capacity information.
    """
    classes = Course.query.all()
    result = []

    for c in classes:
        teacher = User.query.get(c.teacher_id)
        count = Enrollment.query.filter_by(course_id=c.id).count()

        result.append({
            "id": c.id,
            "name": c.name,
            "teacher": teacher.name if teacher else None,
            "time": c.time,
            "capacity": c.capacity,
            "students": count
        })

    return jsonify(result)


# -------------------- Get Student's Enrolled Classes --------------------
@student_bp.route("/myclasses/<int:student_id>", methods=["GET"])
def get_my_classes(student_id):
    """
    Returns all courses in which the specified student is enrolled.
    Includes course details, teacher name, grade, capacity,
    and current number of enrolled students.
    """
    enrollments = Enrollment.query.filter_by(student_id=student_id).all()
    result = []

    for e in enrollments:
        c = Course.query.get(e.course_id)
        teacher = User.query.get(c.teacher_id)
        count = Enrollment.query.filter_by(course_id=c.id).count()

        result.append({
            "id": c.id,
            "class": c.name,
            "teacher": teacher.name,
            "time": c.time,
            "grade": e.grade,
            "students": count,
            "capacity": c.capacity
        })

    return jsonify(result)


# -------------------- Student Sign Up for a Course --------------------
@student_bp.route("/signup", methods=["POST"])
def signup_class():
    """
    Enrolls a student in a selected course.
    Prevents duplicate enrollment and checks whether the class is full.
    Expects JSON with student_id and class_id.
    """
    data = request.json
    student_id = data.get("student_id")
    course_id = data.get("class_id")  # Using frontend naming convention.

    c = Course.query.get(course_id)
    count = Enrollment.query.filter_by(course_id=course_id).count()

    if count >= c.capacity:
        return jsonify({"error": "Class is full"}), 400

    exist = Enrollment.query.filter_by(student_id=student_id, course_id=course_id).first()
    if exist:
        return jsonify({"error": "Already enrolled"}), 400

    new_enroll = Enrollment(student_id=student_id, course_id=course_id, grade=None)
    db.session.add(new_enroll)
    db.session.commit()

    return jsonify({"message": "Enrolled successfully!"})


# -------------------- Student Drop a Course --------------------
@student_bp.route("/drop", methods=["DELETE"])
def drop_class():
    """
    Removes a student from a course they are currently enrolled in.
    Expects student_id and class_id in the JSON body.
    Returns an error if the student is not enrolled in the class.
    """
    data = request.json
    student_id = data.get("student_id")
    course_id = data.get("class_id")

    enrollment = Enrollment.query.filter_by(student_id=student_id, course_id=course_id).first()
    if not enrollment:
        return jsonify({"error": "Not enrolled in this class"}), 404

    db.session.delete(enrollment)
    db.session.commit()

    return jsonify({"message": "Dropped successfully"})
